### PR TITLE
✨ Implement library for topology mutation hooks implementation

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR moves some of the utilities from our test extension into a library that every Runtime extension author can use.

It also adds comments to the test extension so all the implementation choices are explained and the  Runtime extension authors can use it as a reference implementation for their own extension

**Which issue(s) this PR fixes**:
Fixes #

### CODE REVIEW

File: internal/runtime/client/client.go [hunk 1]
```diff
@@ -336,7 +336,7 @@ func (c *client) CallExtension(ctx context.Context, hook runtimecatalog.Hook, fo
 	}
 
 	log.Info(fmt.Sprintf("Calling extension handler %q", name))
-	var timeoutDuration time.Duration
+	timeoutDuration := runtimehooksv1.DefaultHandlersTimeoutSeconds * time.Second
```
Comment (fabriziopandini) [inline internal/runtime/client/client.go:339]: This is consistent with what declared in the API

File: internal/runtime/client/client.go [hunk 2]
```diff
@@ -266,13 +264,15 @@ func aggregateSuccessfulResponses(aggregatedResponse runtimehooksv1.ResponseObje
...
-
+	messages := []string{}
 	for _, resp := range responses {
 		aggregatedRetryResponse.SetRetryAfterSeconds(lowestNonZeroRetryAfterSeconds(
 			aggregatedRetryResponse.GetRetryAfterSeconds(),
 			resp.(runtimehooksv1.RetryResponseObject).GetRetryAfterSeconds(),
 		))
+		messages = append(messages, resp.GetMessage())
 	}
+	aggregatedResponse.SetMessage(strings.Join(messages, ", "))
```
Comment (fabriziopandini) [inline internal/runtime/client/client.go:275]: I know we want back and forth on this but if the extension's errors do not surface the UX degrades a lot.
So I'm proposing to surface errors, document guidelines for error messages, and take time to implement a protection mechanism for flickering conditions

File: internal/runtime/client/client.go [hunk 3]
```diff
@@ -254,25 +254,25 @@ func (c *client) CallAllExtensions(ctx context.Context, hook runtimecatalog.Hook
...
-	// So let's set those values to avoid keeping values that could have been set by the caller of CallAllExtensions.
-	aggregatedResponse.SetMessage("")
+	// At this point the Status should always be ResponseStatusSuccess.
 	aggregatedResponse.SetStatus(runtimehooksv1.ResponseStatusSuccess)
 
-	aggregatedRetryResponse, ok := aggregatedResponse.(runtimehooksv1.RetryResponseObject)
-	if !ok {
-		// If the aggregated response is not a RetryResponseObject then we're done.
-		return
-	}
```
Comment (ykakarap) [inline internal/runtime/client/client.go:266]: I think this was added for safety reasons. For now we control where `CallAllExtensions` are called and with what arguments. But this was added for safety when client will be imported and used by providers that might start implementing their own hooks.

File: test/extension/handlers/lifecycle/handlers.go [hunk 4]
```diff
@@ -36,121 +41,161 @@ import (
...
 		response.Message = err.Error()
 	}
 }
 
-// DoAfterControlPlaneInitialized implements the AfterControlPlaneInitialized hook.
-func (h *Handler) DoAfterControlPlaneInitialized(ctx context.Context, request *runtimehooksv1.AfterControlPlaneInitializedRequest, response *runtimehooksv1.AfterControlPlaneInitializedResponse) {
+// DoAfterControlPlaneInitialized implements the HandlerFunc for the AfterControlPlaneInitialized hook.
+// The hook answer with the response stored in a well know config map, thus allowing E2E tests to
+// control the hooks behaviour during a test.
+// NOTE: custom RuntimeExtension, must implement the body of this func according to the specific use case.
```
Comment (ykakarap) [inline test/extension/handlers/lifecycle/handlers.go:113]:
```suggestion
// The hook answers with the response stored in a well know config map, thus allowing E2E tests to
// control the hook behaviour during a test.
// NOTE: custom RuntimeExtension, must implement the body of this func according to the specific use case.
```

not be consistent with other comments.

File: internal/runtime/client/client.go (hunk 1 above)
Comment (fabriziopandini) [reply internal/runtime/client/client.go:339]: fixed line 622
I think we should left the defaulting here even if it is redundant (thinking to clusterctl alpha plan, where we are going to use a "hacky version" if discovery)

File: exp/runtime/server/server.go [hunk 5]
```diff
@@ -118,9 +125,11 @@ type ExtensionHandler struct {
 	HandlerFunc runtimecatalog.Hook
 
 	// TimeoutSeconds is the timeout of the extension handler.
+	// This is defaulted to 10 seconds if left undefined.
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply exp/runtime/server/server.go:128]: rephrased, but left the value to keep the comment meaningful

File: internal/runtime/client/client.go (hunk 3 above)
Comment (fabriziopandini) [reply internal/runtime/client/client.go:266]: should be fixed now

File: test/extension/handlers/lifecycle/handlers.go [hunk 6]
```diff
@@ -36,121 +41,161 @@ import (
...
+// is to expose HandlerFunc with the signature defined in sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1.
+type Manager struct {
+	client client.Client
 }
 
-// DoBeforeClusterCreate implements the BeforeClusterCreate hook.
-func (h *Handler) DoBeforeClusterCreate(ctx context.Context, request *runtimehooksv1.BeforeClusterCreateRequest, response *runtimehooksv1.BeforeClusterCreateResponse) {
+// NewManager returns a Manager for the lifecycle hooks handlers.
+func NewManager(scheme *runtime.Scheme) (*Manager, error) {
+	restConfig, err := ctrl.GetConfig()
```
Comment (sbueringer) [inline test/extension/handlers/lifecycle/handlers.go:56]: Should we really create our own client inside of this manager? This means we can't share caching etc across the entire RuntimeExtension (compare with creating our own client inside of a webhook or a controller in a regular capi controller)

File: test/extension/handlers/topologymutation/handler.go [hunk 7]
```diff
@@ -220,14 +238,6 @@ func patchKubeadmConfigTemplate(ctx context.Context, k *bootstrapv1.KubeadmConfi
 		}
 		if version.Compare(machineDeploymentVersion, cgroupDriverPatchVersionCeiling) == -1 {
 			log.Info(fmt.Sprintf("Setting KubeadmConfigTemplate cgroup-driver to %q", cgroupDriverCgroupfs))
-			// Set the cgroupDriver in the InitConfiguration.
```
Comment (sbueringer) [inline test/extension/handlers/topologymutation/handler.go:223]: Was this intentionally dropped?

File: exp/runtime/topologymutation/variables.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
+	patchvariables "sigs.k8s.io/cluster-api/internal/controllers/topology/cluster/patches/variables"
+)
+
+// TODO: Add func for validating received variables are of the expected types
+// TODO: Add func for converting complex variables into go structs and/or other basic types.
+
+// GetTemplateVariable get the variable value.
+func GetTemplateVariable(templateVariables map[string]apiextensionsv1.JSON, variableName string) (*apiextensionsv1.JSON, bool, error) {
```
Comment (sbueringer) [inline exp/runtime/topologymutation/variables.go:31]: Should we rename those funcs to GetVariable and GetStringVariable?

I think we don't need template (and the variables could be top-level or from the template)

File: exp/runtime/topologymutation/walker.go [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+type WalkTemplatesOption interface {
+	// ApplyToWalkTemplates applies this configuration to the given WalkTemplatesOptions.
+	ApplyToWalkTemplates(*WalkTemplatesOptions)
+}
+
+// WalkTemplatesOptions contains options for WalkTemplates behavior.
+type WalkTemplatesOptions struct {
+	failForUnknownTypes bool
+	patchFormat         runtimehooksv1.PatchType
+	// TODO: add the possibility to set patchFormat for single patches only, eg. via a func(requestItem) format.
```
Comment (sbueringer) [inline exp/runtime/topologymutation/walker.go:45]: Q: This would be to have different patch formats for different templates?

Is there a use case where we need this?

Comment (sbueringer) [reply exp/runtime/topologymutation/walker.go:45]: I also wonder what the up/downside is of using one patch format over another. Are there specific use cases where I would want to pick one vs the other?

(I think that's something that a user of this lib would probably ask themselves as well)

**Review by sbueringer — COMMENTED** [review]
Great work

File: test/extension/handlers/topologymutation/handler.go (hunk 7 above)
Comment (fabriziopandini) [reply test/extension/handlers/topologymutation/handler.go:223]: yes, InitConfiguration does not apply to worker machines

File: exp/runtime/topologymutation/walker.go (hunk 9 above)
Comment (fabriziopandini) [reply exp/runtime/topologymutation/walker.go:45]: AFAIK there is a couple of differences, mostly about how fields of type array are managed; with merge patch, you have to rebuild the entire array, while with json patch you can append or prepend and probably more.

What I was thinking is, as a runtime extension author, if I hit some problem due to the above limitations, I would prefer to change the patch format for one single patch, not for all (containing the blast radius)

File: test/extension/handlers/topologymutation/handler.go (hunk 7 above)
Comment (sbueringer) [reply test/extension/handlers/topologymutation/handler.go:223]: Ahh, good catch!

File: exp/runtime/topologymutation/walker.go (hunk 9 above)
Comment (sbueringer) [reply exp/runtime/topologymutation/walker.go:45]: > AFAIK there is a couple of differences, mostly about how fields of type array are managed; with merge patch, you have to rebuild the entire array, while with json patch you can append or prepend and probably more.

Yup agree.

Just wonder if it's relevant as the patch will be generated automatically anyway and I would hope that both always generate the right patch. And as nobody has to write the patch manually I would assume nobody really cares how the patch exactly looks like as long as applying the patch leads to the expected result.

> What I was thinking is, as a runtime extension author, if I hit some problem due to the above limitations, I would prefer to change the patch format for one single patch, not for all (containing the blast radius)

I'm not sure, but I suspect both formats can achieve the same kind of modifications they just have to include more or less of the patched resources (but this will be done automatically anyway). I think we should explore this further in a follow-up if we think that one of the formats doesn't work correctly in edge cases. Those would be very hard to catch bugs.

File: test/extension/handlers/lifecycle/handlers.go [hunk 10]
```diff
@@ -36,121 +41,161 @@ import (
...
-// Handler is the handler for the lifecycle hooks.
-type Handler struct {
-	Client client.Client
+// Manager is the Manager for the lifecycle hooks handlers;
+// it provides a common struct shared across all the lifecycle hook handlers because all of them
+// are using a controller runtime client and the same set of func to work with the config map where
+// preloaded answers for lifecycle hooks are stored.
+// NOTE: it is not mandatory to use a Manager in custom RuntimeExtension, what is important
+// is to expose HandlerFunc with the signature defined in sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1.
+type Manager struct {
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply test/extension/handlers/lifecycle/handlers.go:50]: Maybe we should just call it `ExtensionHandlers`.

It's essentially a collection of extension handlers or a wrapper around multiple extension handlers. (we already have funcs like `webhookServer.AddExtensionHandler` and are using this wording in other parts of our code base and in our documentation as well)

File: exp/runtime/topologymutation/walker.go (hunk 9 above)
Comment (fabriziopandini) [reply exp/runtime/topologymutation/walker.go:45]: agreed, it is just a TODO for now

File: test/extension/handlers/lifecycle/handlers.go (hunk 6 above)
Comment (fabriziopandini) [reply test/extension/handlers/lifecycle/handlers.go]: Fixed

File: test/extension/main.go [hunk 11]
```diff
@@ -14,14 +14,18 @@ See the License for the specific language governing permissions and
...
+// - to provide a reference implementation of Runtime Extension
+// - to implement the Runtime Extension used by Cluster API E2E tests.
 package main
 
 import (
 	"flag"
 	"fmt"
 	"net/http"
 	"os"
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
Comment (sbueringer) [inline test/extension/main.go:28]: This should be a linter finding (probably is?)

File: docs/book/src/tasks/experimental-features/runtime-sdk/implement-extensions.md [hunk 12]
```diff
@@ -67,15 +67,17 @@ var (
 )
 
 func init() {
-	// Register the Runtime Hook types into the catalog.
+    // Adds to the catalog all the RuntimeHooks defined in cluster API.
```
Comment (chrischdi) [inline docs/book/src/tasks/experimental-features/runtime-sdk/implement-extensions.md:70]: tabs vs spaces!? also for the other lines?

Seems like all existing lines use tabs, the changed ones now use 4 spaces instead 🤔

Comment (sbueringer) [reply docs/book/src/tasks/experimental-features/runtime-sdk/implement-extensions.md:70]: Good catch!

Let's use tabs consistently like in Go, otherwise the rendered formatting is off

**Comment by sbueringer** [comment]
lgtm pending https://github.com/kubernetes-sigs/cluster-api/pull/7413#discussion_r1018000946 + squash

File: docs/book/src/tasks/experimental-features/runtime-sdk/implement-extensions.md (hunk 12 above)
Comment (fabriziopandini) [reply docs/book/src/tasks/experimental-features/runtime-sdk/implement-extensions.md]: fixed, nice catch

**Comment by sbueringer** [comment]
Thx!

/lgtm
/approve
